### ISSUE TITLE
hexchat system tray icon missing in kde5 plasma next

### ISSUE THREAD

**Issue by Swoorup** [hexchat/hexchat#1235]
hexchat linux fails to register system tray icon in kde plasma version 5.4.0

**Comment by KaiRo-at** [comment]
The issue probably is that plasma 5 does nut support the old XEmbed tray spec any more and tray icons need to use "status notifier" (also known as "application indicator" from the Unity world) as provided by libindicator/libappindicator in the GTK+ world.

**Comment by TingPing** [comment]
@KaiRo-at Yup, see https://github.com/hexchat/hexchat/tree/wip/appindicator

**Comment by KaiRo-at** [comment]
Awesome! \o/ Is this coming in a soon-upcoming version?

**Comment by TingPing** [comment]
Well there are a few issues keeping it from being merged:

- Appindicators use middle click for actions (aka hide/show hexchat)
- They rely upon icon files and can't be assigned an image directly

The former will just cause people complaining to me for the next year.
The later may have some performance issues, HexChat defaults to 'blinking' the icon often which is stupid but again if behavior changes nobody shuts up about it.

**Comment by KaiRo-at** [comment]
Well, not supporting app indicator is causing complaints from early switchers to Plasma 5, and once distros start switching to it by default (openSUSE will for the next release, it seems, others will follow) then those complaints will start to get much louder. I guess Unity users are not complaining because they never saw it actually working, but they are missing the feature completely.

The blinking is the reason why I minimized the cases where I get notified, as blinking stuff is really distracting when you are actually doing work - but that's just my opinion (I'd prefer a different static icon with color instead of the blinking).

I'd say that in the end, not having the indicator at all is worse than a changed behavior.

All that said, is there some way to either switch this automatically or via pref? It would really be good if something could be available to those using a modern desktop environment.

**Comment by Mikaela** [comment]
If I understood correctly, you can get the old icons in Plasma 5.

* https://wiki.archlinux.org/index.php/KDE#System_tray_icons
* http://blog.martin-graesslin.com/blog/2014/06/where-are-my-systray-icons/

**Comment by KaiRo-at** [comment]
I have all the indicator packages and stuff installed and it doesn't work. You only get the old icons when app indicator is supported.

**Comment by karolherbst** [comment]
@TingPing: doesn't the "middle" click part depend on the sessionManager? I know, that the events can be "overwritten" in the client application, but if this is implemented good, then the session manager will take care of theming, open/close and context menu. Otherwise I can't explain, why gtk application usually have my Qt5 theme (and not Qt4 or gtk theme) in the tray.

**Comment by TingPing** [comment]
@karolherbst I can't say I've read the spec or anything. I know for a fact though that the fallback icon is forced to use middle click, so users of xfce, gnome, etc will certainly be complaining.

**Comment by karolherbst** [comment]
I see

**Comment by est31** [comment]
Issue for kubuntu [here](https://bugs.launchpad.net/ubuntu/+source/hexchat/+bug/1455279). Created that, because [this](http://blog.martin-graesslin.com/blog/2014/06/where-are-my-systray-icons/) blog post said its only about build dependencies .

**Comment by TingPing** [comment]
> because this blog post said its only about build dependencies .

Its about actually using those libraries, not a simple build change. Anyway the branch linked above uses
libappindicator.

**Comment by simotek** [comment]
For what its worth the next enlightenment release will also only support app indicator, the last release e19 strongly discourages xembed. The xembed protocol was always very hard for non gtk window managers to impement well. 

@est31 that blog post is talking about Qt4 apps where it is only a build change.

@TingPing in enlightenment the menu is shown on left click, I believe whats used to show the window unless your talking about fallback for non appindicator systray.

From what I've heard gnome is also slowly moving toward app indicator as well, someone told me xembed doesn't really work in wayland although maybe people have come up with hacks around that.

I'll give your branch a try.

Simon Lees

openSUSE - Enlightenment maintainer

**Comment by TingPing** [comment]
>From what I've heard gnome is also slowly moving toward app indicator as well

Gnome wants all system trays to die. They have no interest in any replacements atm.

**Comment by simotek** [comment]
yeah its decisions like that that are the reason I stopped following gnome dev very soon after the gnome 3 launch, there is a extension for it though https://extensions.gnome.org/extension/615/appindicator-support/

**Comment by KaiRo-at** [comment]
As for the Wayland, thing, it definitely does not support XEmbed, which is a X-specific thing.

@simotek I think we should get the openSUSE package manager for the hexchat package to switch to the branch that @TingPing mentioned, at least until the actual master branch will catch up with the current state of reality where XEmbed tray is more a zombie than a living thing.

**Comment by TingPing** [comment]
>@simotek I think we should get the openSUSE package manager for the hexchat package to switch to the branch that @TingPing mentioned

No... that is not how packaging (should) works, they package what developers deem stable.

**Comment by KaiRo-at** [comment]
Well, the official tree is obviously broken in multiple desktops of openSUSE Tumblweed, including the default one, which is Plasma 5.

**Comment by Arnavion** [comment]
Existing installs of Tumbleweed will still stay on kde4 desktop, so it's not as bad as that. They'd have to explicity choose to install plasma5-desktop which will then replace kde4base-workspace.

I guess you're saying the default for *new* installations of Tumbleweed is plasma5-desktop now?

**Comment by Poorchop** [comment]
[KDE Plasma 5 Becomes the Default Desktop of OpenSUSE Tumbleweed](https://news.opensuse.org/2015/05/16/tumbleweed-moves-to-plasma-5-3-and-a-new-release-of-kde-applications/)

**Comment by KaiRo-at** [comment]
Arnavion, KDE4 is dead and not maintained any more. Tumbleweed, Kubuntu, and other distros are *replacing* KDE4 with Plasma 5. Completely. Without any way to go back.

**Comment by Arnavion** [comment]
Weird... Poorchop's link seems to say yast should offer to replace kdebase4-workspace with plasma5-desktop, but that's not what happened with me. I only got the upgrade for *applications* to their kf5 versions. It didn't offer to install plasma5-desktop on its own - I had to explicitly install it.

Anyway, enough thread-derailing. I reverted back to kde4 for the moment until plasma5 is good enough for me (no quicklaunch widget, the automatic QSTI->KSNI doesn't work properly, etc.) so I'm not going to push to have this working in the short term. Whether TingPing finishes it on his own or not is up to him.

**Comment by simotek** [comment]
@KaiRo-at , @TingPing Is right, something non stable shouldn't be packaged into official repos, rather then building from this branch directly it would be nicer if the branch was merged with a --app-indicator configure flag so that from the next release distros / packagers can decided which version they build. It is very rare for people to officially package something from a git branch.

**Comment by TingPing** [comment]
>Well, the official tree is obviously broken in multiple desktops of openSUSE Tumblweed, including the default one, which is Plasma 5.

Upstream DEs made this decision, they absolutely knew many legacy applications would be broken for at least a period of time.

The appindicator work should be merged by the next release whenever that may be. My concerns about changing behavior still stand though. If I force new behavior on users I'll still be getting more complaints than it not working at all...

**Comment by KaiRo-at** [comment]
@TingPing, I then hope that the next release will be soon, as right now not having this is one reason why my system is getting insecure as I can't update (and get Plasma 5 with that) as long as I don't see notifications of being pinged on IRC.

I understand the concerns, it's a tradeoff between users complaining on one side (functionality changing), and complaining or even switching to a different software on the other (things not working). I feel with you there, I'm in discussions about similar tradeoffs every now and then myself (and I saw that in the end, the more modern technology usually wins out).

**Comment by TingPing** [comment]
>@TingPing, I then hope that the next release will be soon, as right now not having this is one reason why my system is getting insecure as I can't update (and get Plasma 5 with that) as long as I don't see notifications of being pinged on IRC.

You are ofc free to build that branch yourself, just don't expect your distro to do so.

**Comment by simotek** [comment]
@KaiRo-at It's what comes with running a rolling release distro, I'm mostly working on trying to have everything working for the next stable release thats all i have time to manage

**Comment by rapgro** [comment]
Not sure if it's related exactly. But this issue here is a GNOME thingy.
https://bugzilla.gnome.org/show_bug.cgi?id=733472

**Comment by TingPing** [comment]
> Not sure if it's related exactly. But this issue here is a GNOME thingy.

Uh, I'm not sure how that is related at all.

**Comment by Monkey0x9** [comment]
FYI: The latest KDE plasma release does support legacy icons, so the HexChat icon is visible again. They mentioned it in the changelog: https://www.kde.org/announcements/plasma-5.5.0.php
I've also tested it, and it works :)

**Comment by Arnavion** [comment]
Cool, thanks!

**Comment by TingPing** [comment]
I still think that app-indicator support is a good idea and I already marked app-indicator issues as dupes of this so I will just re-purpose this towards that.

**Comment by LiamDawe** [comment]
What's happening with this? The 3rd part indicator for us on Ubuntu Unity is buggy and opens a new instance when you click on it.

An official one would be great, so many other apps have them now it's a pain not to have one since Unity doesn't support standard tray icons.

**Comment by 2E0PGS** [comment]
Ubuntu 16.04 with Hexchat 2.10.2. No tray icon. The 3rd party program hexchat-indicator don't work either.

**Comment by kilbith** [comment]
Fix this please, it's annoying.

**Comment by LiamDawe** [comment]
Just to jump in here, that this has become essential. Especially with Ubuntu 17.10 now using Gnome Shell by default, making Hexchat have no tray support in any distribution that ships with the latest Gnome.

**Comment by gtirloni** [comment]
Install the "Topicons plus" extension. It works just fine with Hexchat.

**Comment by LiamDawe** [comment]
Installing yet another extension is not a solution when that replaces the normal tray in Ubuntu.

**Comment by gtirloni** [comment]
Welcome to GNOME :)

This is not an Hexchat issue.

**Comment by LiamDawe** [comment]
Fine. I will just use Quassel which supports indicators, no problem.

**Comment by jhasse** [comment]
> This is not an Hexchat issue.

It is. The legacy tray was deprecated for good reasons.

**Comment by parkerbryan** [comment]
> Welcome to GNOME :)

This is not specific to GNOME.

**Comment by simotek** [comment]
yes it is, kde, enlightenment and every other desktop moving to wayland
have this issue

consider it a hexchat wayland compatibility issue

**Comment by Krenair** [comment]
@gtirloni I tried topicons plus and it just adds a HexChat icon to the top, which doesn't do the old listing of channels with unread notifications functionality.

**Comment by cdysthe** [comment]
Hexchat really needs tray icon support on Ubuntu and Plasma. I mean, really! :)

**Comment by IvanMalison** [comment]
> This is not specific to GNOME.

In fact, gnome is the only major desktop environment that DOESN'T have app indicator support out of the box (You need this extension: https://github.com/ubuntu/gnome-shell-extension-appindicator)

**Comment by Mau-G** [comment]
Is there an update on this issue? The newer xfce4-statusnotifier-plugin uses appindicators.

https://goodies.xfce.org/projects/panel-plugins/xfce4-statusnotifier-plugin

**Comment by blade-z** [comment]
> @KaiRo-at Yup, see https://github.com/hexchat/hexchat/tree/wip/appindicator

This is no longer an option, is there a more current branch I can build that supports this? 

I can't believe it's almost 2019 and this is still an issue

**Comment by Glandos** [comment]
Everyone had a look at https://github.com/kitsunyan/appindicator-collection ?

**Comment by TingPing** [comment]
> Anyone had a look at [kitsunyan/appindicator-collection](https://github.com/kitsunyan/appindicator-collection) ?

Wow, what a joke. The author could have just wrote a hexchat plugin that would have easily been official and shipped with distros but instead wrote a hack that dynamically overrides internal GTK functions...

**Comment by Uzzi** [comment]
https://github.com/hexchat/hexchat/issues/2386

**Comment by RdeWilde** [comment]
So, still no tray icon. Wow...

**Comment by AsciiWolf** [comment]
Any update? I was not able to get tray icon working on GNOME with KStatusNotifierItem/AppIndicator Support extension installed.

**Comment by RdeWilde** [comment]
Just delete the comments, nice. I'll go ahead and use another app, this is shit :wave:

**Comment by simotek** [comment]
In the last couple of weeks I have revived @TingPing 's old branch with support for app-indicator and rebased it against current master you can find it here https://github.com/simotek/hexchat/tree/wip/appindicator (note is still in the `wip/appindicator` branch)

I have cleaned up most of the commented out code blocks and other commented out code as well as fixing some of the bugs and implementing a couple of missing things. It is now at the point where it works well for me which means it would be great if some other people tested it. 

To build with appindicator support pass `-Dwith-appindicator=true` to meson, if you are installing to a prefix other then `/usr/` you will need to copy the `hexchat-tray*` icons from `prefix/share/icons/hicolor/48x48/apps` into `/usr/share/icons/hicolor` or `/usr/share/pixmaps`. 

I am yet to look at the diff between my branch and master so there is likely some bits of the code that should be further cleaned up before it goes for review, i'll likely do that in a couple of weeks before submitting it for review but feedback is welcome now

**Comment by 0ther0ne** [comment]
> feedback is welcome now

![HexChat_appindicator](https://user-images.githubusercontent.com/64803452/81483634-f046fb00-923f-11ea-8826-98bcf124d78a.png)

works for me in Ubuntu 20.04

**Comment by terencode** [comment]
It's beautiful :heart: 

![image](https://user-images.githubusercontent.com/12665563/81483796-25078200-9241-11ea-8039-3afe2ed4ff1c.png)

**Comment by AsciiWolf** [comment]
Amazing! @simotek Thank you a much for working on this!

By the way, I wonder if it could be possible to make this a user-installable plugin for HexChat.

**Comment by simotek** [comment]
> Amazing! @simotek Thank you a much for working on this!
> 
> By the way, I wonder if it could be possible to make this a user-installable plugin for HexChat.

I don't think so, atleast not easily, this code is just piggybacking off the existing tray implementations inside hexchat, I am not much of a gtk / hexchat developer so I don't know the plugin api so it maybe possible but it would probably be a rewrite.

The existing plugin api has a way of letting plugins change the tray icon but I didn't implement it at the moment because it would need a different api. The existing api just passes a the tray a pixmap to display where as appindicator generally passes an fdo icon name to the program running the tray via dbus then that tray program finds and shows the icon.

**Comment by simotek** [comment]
> It's beautiful heart
> 
> Some suggestions:
> 
>     * Implement toggle visibility using double click  (calls “Activate”)

Sure, I can look at this, the tray I use doesn't support Double click so i'll need some help testing.

>     * Add an option in hexchat to "minimize to tray"

The "toggle visibility" option should already do that if the window is showing, is that good enough? most of the apps in my tray have something similar to that (although I don't normally use that functionality, I just keep stuff maxmised on different Virtual Desktops / Workspaces

**Comment by terencode** [comment]
> Sure, I can look at this, the tray I use doesn't support Double click so i'll need some help testing.

Count on me for testing.

About "Add an option in hexchat to "minimize to tray", if you can implement my first request it might be overkill. I was thinking about it because many apps have this option so that if you click on the "close" or alt+f4, you don't actually quit the app but minimizes it. (e.g. Discord, MellowPlayer, Steam,  Audacious. Psensors, Riot..). Although for most of them it's not enabled by default.

**Comment by simotek** [comment]
> > Sure, I can look at this, the tray I use doesn't support Double click so i'll need some help testing.
> 
> Count on me for testing.
> 
> About "Add an option in hexchat to "minimize to tray", if you can implement my first request it might be overkill. I was thinking about it because many apps have this option so that if you click on the "close" or alt+f4, you don't actually quit the app but minimizes it. (e.g. Discord, MellowPlayer, Steam, Audacious. Psensors, Riot..). Although for most of them it's not enabled by default.

Yeah it certainly seems like "Toggle Visibility" is already doing just that, Rocket Chat is slightly smarter and changes the menu option from Show to Hide

**Comment by zx2c4** [comment]
@simotek Any plans for merging this to master?
